core/tests/puppeteer-acceptance-tests/functions/is-element-clickable.ts

```diff
@@ -0,0 +1,170 @@
+   * the target element.
+   *
+   * @param {Element} targetElement - The element to compare with the other
+   * overlapping elements.
+   * @param {Element[]} overlappingElements - The overlapping elements to compare
+   * against the target element.
+   */
+  const isOverlappingElementsMatching = (
+    targetElement: Element,
+    overlappingElements: Element[]
+  ): boolean => {
+    if (
+      overlappingElements.some(
+        el => el === targetElement || targetElement.contains(el)
+      )
+    ) {
+      return true;
+    }
+
+    let elementsWithShadow = [...new Set(overlappingElements)];
+    elementsWithShadow = elementsWithShadow.filter(
+      el => el && el.shadowRoot && el.shadowRoot.elementFromPoint
+    );
+
+    let shadowElements: Element[] = [];
+    for (const shadowEl of elementsWithShadow) {
+      if (shadowEl.shadowRoot) {
+        shadowElements.push(
+          ...getOverlappingElements(targetElement, shadowEl.shadowRoot)
+        );
+      }
+    }
+    shadowElements = [...new Set(shadowElements)];
+    shadowElements = shadowElements.filter(
+      el => !overlappingElements.includes(el)
+    );
+
+    if (shadowElements.length === 0) {
+      return false;
+    }
+
+    return isOverlappingElementsMatching(targetElement, shadowElements);
+  };
+
+  /**
+   * This function checks if the given element is in the viewport.
+   */
+  const isElementInViewport = (element: Element): boolean => {
+    const elementDimensions = element.getBoundingClientRect();
+
+    const windowHeight =
+      window.innerHeight || document.documentElement.clientHeight;
+    const windowWidth =
+      window.innerWidth || document.documentElement.clientWidth;
+
+    const verticalInView =
+      elementDimensions.top <= windowHeight &&
+      elementDimensions.top + elementDimensions.height > 0;
+    const horizontalInView =
+      elementDimensions.left <= windowWidth &&
+      elementDimensions.left + elementDimensions.width > 0;
+
+    return verticalInView && horizontalInView;
+  };
+
+  /**
+   * This function checks if the element is clickable, by checking if it is
+   * in the viewport and not blocked by any other element. This function also
+   * checks if the element is disabled, applicable for elements like buttons
+   * and form elements.
+   */
+  const isClickable = (element: Element): boolean => {
+    return (
+      (element as HTMLFormElement).disabled !== true &&
+      isElementInViewport(element) &&
+      isOverlappingElementsMatching(element, getOverlappingElements(element))
+    );
+  };
+
+  // Here we check if the element is clickable and if not, we scroll it into view
+  // and check again. We do this twice to ensure that the element is centered and
+  // not blocked.
+  if (!isClickable(element)) {
+    element.scrollIntoView({block: 'center', inline: 'center'});
+
+    if (!isClickable(element)) {
+      element.scrollIntoView({block: 'center', inline: 'center'});
```

Comment:
Why would doing this twice ever have different results from doing it once?

Reply:
Removed